docs/proposals/templates.md

```diff
@@ -0,0 +1,404 @@
+process templates.*
+
+* Parameter definitions include metadata which describes the purpose of the parameter.  Since parameter definitions are part of the template,
+  there is no need to indicate which template they correspond to.
+
+*It should be possible to validate templates and template parameters, both values and the schema.*
+
+* Template objects are subject to standard api validation.
+
+*It should also be possible to validate and view the output of the substitution process.*
+
+* The `/processedTemplates` api returns the result of the substitution process, which is itself a Template object that can be validated.
+
+*It should be possible to generate forms for parameterized templates, as discussed in #4210 and #6487.*
+
+* Parameter definitions provide metadata that allows for the construction of form-based UIs to gather parameter values from users.
+
+*It shouldn't be inordinately difficult to evolve templates. Thus, strategies such as versioning and encapsulation should be 
+encouraged, at least by convention.*
+
+* Templates can be versioned via annotations on the template object.
+
+## Alternatives
+
+### Mustache other generic templating language
+
+[Mustache](https://mustache.github.io/) offers a semi-standard generic way to do templating with substitution.  The primary 
+limitation is that the syntax is not valid json (unless all parameters are strings and delimited by quotes), and parameter 
+values must be provided separately.  This means Mustache templates could not be stored and referenced as first class objects.
+
+Other templating systems surveyed had similar limitations:  separated parameters from the template, did not provide generator 
+capabilities, and/or did not use a syntax which could easily be rendered as a Kubernetes API object.
+
+#### Stream transformers
+As defined [here](https://github.com/kubernetes/kubernetes/pull/14993) and [here](https://github.com/kubernetes/kubernetes/pull/14918).
+
+The primary limitation here seems to be that it is a client-side proposal, generators would have to be provided in the client, meaning third party clients/web consoles/etc would have a hard time taking advantage of this mechanism.  It also doesn’t provide a way to create and store new generic templates, you must have a generator defined and then provide appropriate parameters for that generator, so it doesn’t seem like it fills the same space as the templates in this proposal.
+
+Lastly it doesn’t provide a way to list/describe parameters consumed by the template, or provide default values in the template itself, making it hard to build a UI experience around.
+
+#### Deployment Manager
+
+As described [here](https://github.com/kubernetes/deployment-manager)
+
+* Utilizes multiple files with different syntax to define an application topology.
+* Does not define a first class API object.
+* Does not appear to natively offer generation logic for producing parameter values.
+
```

Comment:
> Also, there's no requirement to use both languages in a single example, or to use multiple files, so both assertions are inaccurate.

is it possible to write a single file, in k8s-compatible json syntax, which provides parameterization?  That is what is achieved by the openshift template schema, and what I am trying to distinguish here as a desirable trait.  Whether an example can be created which doesn't need additional syntax or files isn't really my point, the question is whether the full power of DM can be realized without branching out into multiple files and multiple syntaxes (above and beyond the k8s syntax).  

My impression is that the answer is no.  Is it a desirable goal (single file, k8s syntax, api object)?  in my opinion, based on the user experience we have been able to deliver in openshift by integrating templates as a first class concept, is that it is, but I think we'll be hashing that out in the syntax/schema discussion once I update this PR per our discussion yesterday.

Regarding parameter generation, I agree with the concerns about it being compiled into the runtime and it would be nice to make it more flexible.  On the other hand, by being a little opinionated here, we provided a generation function that covers probably 70% of the use cases for parameter generation (no one has come to us so far asking for additional generation functionality) in an extremely consumable format that minimizes complexity or requirements to learn new languages.